flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java

```diff
@@ -282,18 +282,29 @@ private void isReadyForDeployment(ClusterSpecification clusterSpecification) thr
 		}
 
 		// Check if we don't exceed YARN's maximum virtual cores.
-		// The number of cores can be configured in the config.
-		// If not configured, it is set to the number of task slots
-		int numYarnVcores = yarnConfiguration.getInt(YarnConfiguration.NM_VCORES, YarnConfiguration.DEFAULT_NM_VCORES);
+		// Fetch numYarnMaxVcores from all the RUNNING nodes via yarnClient
+		int numYarnMaxVcores = Integer.MIN_VALUE;
```

Comment:
How about:
```
		final int numYarnMaxVcores;
		try {
			numYarnMaxVcores = yarnClient.getNodeReports(NodeState.RUNNING)
				.stream()
				.mapToInt(report -> report.getCapability().getVirtualCores())
				.max()
				.orElse(0);
		} catch (Exception e) {
			throw new YarnDeploymentException("Couldn't get cluster description, please check on the YarnConfiguration", e);
		}
```
?

Pros of doing so:
- `numYarnMaxVcores` is final
- If no NodeManagers are running at all, we would log `0` vcores instead of `Integer.MIN_VALUE`
- Concise & easy to read

Reply:
this way sure better than mine. I will do the update, thanks @GJL

Reply:
update done and passed local verification